### ISSUE TITLE
关于英文BART预训练

### ISSUE THREAD

**Issue by PALIN2018** [dbiir/UER-py#241]
首先感谢您提供优质的代码！
如题，我想使用您提供的代码实现英文BART的预训练，需要找您确认以下事项：
1）数据预处理的格式
项目提供的样例如下：
<img width="949" alt="image" src="https://user-images.githubusercontent.com/35458200/146870805-1cc0ab64-6ecd-430f-8657-e4515903fc81.png">
所以英文应该采用相同的格式，即每句话占一行，每段文本之间用空行隔开。
2）数据预处理及模型预训练的分词器的选择
在处理英文文本时，直接使用默认的BertTokenizer就可以了吗？还是要使用XLM-Roberta的分词器？
如果我要在Facebook release的BART-base上继续预训练，是应该使用该预训练模型的BARTTokenizer还是可以继续用项目提供的Tokenizer即可？
3）词表文件vocab
由于我使用的是Facebook release的BART-base模型，并基于该模型继续预训练，所以需要把vocab.json的文件转换为txt。
4）preprocess和pretrain超参数设置
请问做英文预训练时，这些超参数的设置和中文有什么差异么？还是直接使用提供的中文训练的样例就可以？有什么需要注意的地方嘛？

提问有点多，希望您有空可以回答，谢谢！

**Comment by hhou435** [comment]
1.是的，使用相同格式即可
2.分词器选择和原模型保持一致即可
我看了Facebook 的BART-base应该是用了bpe tokenzier
您需要在预处理和预训练时指定--tokenizer bpe --vocab_path models/huggingface_gpt2_vocab.txt --merges_path models/huggingface_gpt2_merges.txt
vocab_path和merges_path对应huggingface仓库中的vocab.json和merges.txt
vocab.json需要转换成对应的txt文件
3.同上
4.可以先尝试使用中文样例中的配置，观察一下效果

**Comment by PALIN2018** [comment]
感谢您的回复！我目前使用单机多卡进行预训练，但出现了严重的负载不均衡的问题，请问您是否有解决思路呢
![37571640691140_ pic_hd](https://user-images.githubusercontent.com/35458200/147572268-66638c60-c49e-4d30-ba5c-889a2d2eef7e.jpg)
？

**Comment by Abigial61** [comment]
您好~  想请问一下在指定bpe tokenzier的过程中，好像代码有报错，preprocess.py: error: argument --tokenizer: invalid choice: 'bpe' (choose from 'bert', 'char', 'space', 'xlmroberta')，模型中不存在bpe，是需要像分词与词典wiki文档里的"
通过 --spm_model_path 指定sentencepiece模型路径，然后导入sentencepiece模块，加载sentencepiece模型，对句子进行切分"来指定另外的tokenizer吗？如果想要使用huggingface transformer的tokenizer模型的话该怎么导入呢？

**Comment by hhou435** [comment]
> @hhou435 感谢您的回复！我目前使用单机多卡进行预训练，但出现了严重的负载不均衡的问题，请问您是否有解决思路呢？谢谢！ 加载的是Facebook 的bart-base，batchsize=24，seq_length=128
> 
> ![37571640691140_ pic_hd](https://user-images.githubusercontent.com/35458200/147572268-66638c60-c49e-4d30-ba5c-889a2d2eef7e.jpg)

您好，我们训练BART的时候没出现这个问题，您可以提供一下训练命令，我们尝试复现一下

**Comment by hhou435** [comment]
> 您好~ 想根据您的回答补充几个问题，还烦请指教 1.请问一下在指定bpe tokenzier的过程中，好像代码有报错，模型中不存在bpe。 preprocess.py: error: argument --tokenizer: invalid choice: 'bpe' (choose from 'bert', 'char', 'space', 'xlmroberta')，
> 
> 是需要像wiki文档里指出的使用另外指定的tokenizer，"通过 --spm_model_path 指定sentencepiece模型路径，然后导入sentencepiece模块，加载sentencepiece模型，对句子进行切分"来指定另外的tokenizer吗？ 如果是的话，想要使用huggingface transformer的tokenizer模型的话该怎么导入呢？
> 
> 2. --merges_path参数似乎在preprocess和pretrain中都不存在，该怎么办呢？

您好，需要下载最新版本才可以使用bpe tokenizer

**Comment by PALIN2018** [comment]
> > @hhou435 感谢您的回复！我目前使用单机多卡进行预训练，但出现了严重的负载不均衡的问题，请问您是否有解决思路呢？谢谢！ 加载的是Facebook 的bart-base，batchsize=24，seq_length=128
> > ![37571640691140_ pic_hd](https://user-images.githubusercontent.com/35458200/147572268-66638c60-c49e-4d30-ba5c-889a2d2eef7e.jpg)
> 
> 您好，我们训练BART的时候没出现这个问题，您可以提供一下训练命令，我们尝试复现一下
您好，祝您新年快乐！我们预训练BART的采用的是增量预训练的方式，即先使用您提供的转换代码将hugging face上的facebook 提供的预训练好的bart-base模型转换为uer格式，然后使用如下指令进行预训练，您复现可以尝试这种增量预训练的方式。
python pretrain.py --dataset_path ${model_name}pretrain_10000.pt 
--tokenizer bpe 
--merges_path huggingface_pretrained_models/${model_name}/merges.txt 
--total_steps ${total_steps} 
--save_checkpoint_steps 1000 
--config_path models/${model_type}/${model_name}config.json 
--vocab_path huggingface_pretrained_models/${model_name}/vocab.txt 
--pretrained_model_path models/${model_type}/${model_name}.bin 
--output_model_path models/${model_type}/tweet${model_type}${model_size}model.bin 
--world_size 7 --gpu_ranks 0 1 2 3 4 5 6 --learning_rate 2e-5 
--batch_size ${batch_size} 
--embedding word_pos_seg --encoder transformer --mask fully_visible --target mlm > exp_log/m${model_name}bs${batch_size}ts${total_steps}_multi_gpu.log 2>&1 非常感谢
